File: flower_ui/src/models/product.ts
/**
 * Product model types and Zod validation schemas for the flower shop application
 * Matches the backend Product entity structure
 */

import { z } from 'zod';

// ====== Zod Validation Schemas ======

export const productSearchSchema = z.object({
  keyword: z.string().optional(),
  categoryId: z.number().positive().optional(),
  status: z.union([z.literal(0), z.literal(1)]).optional(),
  featured: z.union([z.literal(0), z.literal(1)]).optional(),
  minPrice: z.number().min(0).optional(),
  maxPrice: z.number().min(0).optional(),
  stockStatus: z.enum(['in_stock', 'low_stock', 'out_of_stock']).optional(),
  sortBy: z.enum(['created_at', 'price', 'name', 'stock_quantity', 'sales']).default('created_at'),
  sortOrder: z.enum(['asc', 'desc']).default('desc'),
  current: z.number().min(1).default(1),
  size: z.number().min(1).default(12),
});

export const productFormSchema = z.object({
  name: z.string().min(1, '商品名称不能为空').max(100, '商品名称不能超过100个字符'),
  description: z.string().optional(),
  price: z.number().min(0.01, '价格必须大于0'),
  originalPrice: z.number().min(0).optional(),
  images: z.array(z.instanceof(File)).optional(),
  flowerLanguage: z.string().optional(),
  careGuide: z.string().optional(),
  categoryId: z.number().positive('请选择商品分类'),
  specification: z.string().optional(),
  status: z.union([z.literal(0), z.literal(1)]).default(1),
  featured: z.union([z.literal(0), z.literal(1)]).default(0),
  stockQuantity: z.number().min(0, '库存数量不能小于0').default(0),
  lowStockThreshold: z.number().min(0, '库存预警阈值不能小于0').default(5),
}).refine(
  (data) => {
    if (data.originalPrice === undefined) return true;
    return data.price <= data.originalPrice;
  },
  {
    message: '售价不能高于原价',
    path: ['price'],
  }
);

export const productUpdateSchema = productFormSchema.partial().extend({
  id: z.number().positive('商品ID必须大于0'),
});

// ====== TypeScript Types (derived from Zod schemas) ======

export interface Product {
  id: number;
  name: string;
  description: string;
  price: number;
  originalPrice?: number;
  images?: any; // 旧字段，兼容性保留，可能是JSON字符串或对象数组
  imageList?: string[]; // 新的图片列表
  mainImagePath?: string; // 主图路径
  flowerLanguage?: string;
  careGuide?: string;
  categoryId: number;
  categoryName?: string; // from category join
  specification?: string;
  status: 0 | 1; // 0-下架，1-上架
  featured: 0 | 1; // 0-不推荐，1-推荐
  stockQuantity: number;
  lowStockThreshold: number;
  createdAt: string;
  updatedAt: string;
}

export type ProductFormData = z.infer<typeof productFormSchema>;
export type ProductUpdateData = z.infer<typeof productUpdateSchema>;
export type ProductFilters = z.infer<typeof productSearchSchema>;

export interface ProductApiResponse {
  records: Product[];
  total: number;
  size: number;
  current: number;
  pages: number;
}

export interface ProductImage {
  url: string;
  file: File;
  preview?: string;
}

export interface ProductImages {
  main?: string;
  subImages: string[];
}

// ====== Utility Functions ======

// Stock status utilities
export const getStockStatus = (product: Product): 'in_stock' | 'low_stock' | 'out_of_stock' => {
  if (product.stockQuantity <= 0) return 'out_of_stock';
  if (product.stockQuantity <= product.lowStockThreshold) return 'low_stock';
  return 'in_stock';
};

export const getStockStatusText = (product: Product): string => {
  const status = getStockStatus(product);
  switch (status) {
    case 'in_stock': return '库存充足';
    case 'low_stock': return '库存不足';
    case 'out_of_stock': return '缺货';
    default: return '未知';
  }
};

export const getStockStatusColor = (product: Product): 'success' | 'warning' | 'error' | 'default' => {
  const status = getStockStatus(product);
  switch (status) {
    case 'in_stock': return 'success';
    case 'low_stock': return 'warning';
    case 'out_of_stock': return 'error';
    default: return 'error'; // Fallback to error or handle appropriately, but 'default' is not in type
  }
};

export const hasDiscount = (product: Product): boolean => {
  return product.originalPrice != null && product.originalPrice > product.price;
};

export const getDiscountPercentage = (product: Product): number => {
  if (!hasDiscount(product)) return 0;
  return Math.round(((product.originalPrice! - product.price) / product.originalPrice!) * 100);
};

// Status helpers
export const getProductStatusText = (status: 0 | 1): string => {
  return status === 1 ? '上架' : '下架';
};

export const getFeaturedText = (featured: 0 | 1): string => {
  return featured === 1 ? '推荐' : '普通';
};

export const getProductStatusColor = (status: 0 | 1): 'success' | 'error' => {
  return status === 1 ? 'success' : 'error';
};

// Image utilities
export const parseImagesJson = (imagesJson: string): string[] => {
  try {
    return JSON.parse(imagesJson || '[]');
  } catch {
    return [];
  }
};

export const imagesToJson = (images: string[]): string => {
  return JSON.stringify(images);
};

// New: Parse ProductImages structure
export const parseProductImages = (imagesJson: string): ProductImages => {
  try {
    if (!imagesJson || imagesJson.trim() === '') {
      return { subImages: [] };
    }

    const parsed = JSON.parse(imagesJson);

    // Check if it's already the new format with main and subImages
    if (parsed && typeof parsed === 'object' && ('main' in parsed || 'subImages' in parsed)) {
      return {
        main: parsed.main || undefined,
        subImages: parsed.subImages || []
      };
    }

    // Try to parse as array (legacy format)
    if (Array.isArray(parsed)) {
      if (parsed.length > 0) {
        return {
          main: parsed[0],
          subImages: parsed.slice(1)
        };
      } else {
        return { subImages: [] };
      }
    }

    return { subImages: [] };
  } catch (error) {
    console.warn('Failed to parse product images JSON:', imagesJson, error);
    return { subImages: [] };
  }
};

// Convert ProductImages to JSON string
export const productImagesToJson = (productImages: ProductImages): string => {
  return JSON.stringify(productImages);
};

// Get all images (main + subImages)
export const getAllProductImages = (productImages: ProductImages): string[] => {
  const allImages: string[] = [];
  if (productImages.main) {
    allImages.push(productImages.main);
  }
  allImages.push(...productImages.subImages);
  return allImages;
};

// Get main image or fallback to first image
export const getMainProductImage = (imagesJson: string): string | undefined => {
  const productImages = parseProductImages(imagesJson);
  return productImages.main || productImages.subImages[0];
};

// Default form values
export const getDefaultProductFormData = (): ProductFormData => ({
  name: '',
  description: '',
  price: 0,
  originalPrice: undefined,
  images: [],
  flowerLanguage: '',
  careGuide: '',
  categoryId: 0,
  specification: '',
  status: 1,
  featured: 0,
  stockQuantity: 0,
  lowStockThreshold: 5,
});

// Form helper functions
export const prepareProductFormData = (product?: Partial<Product>): ProductFormData => {
  const defaultData = getDefaultProductFormData();

  if (!product) return defaultData;

  return {
    ...defaultData,
    name: product.name || '',
    description: product.description || '',
    price: product.price || 0,
    originalPrice: product.originalPrice,
    images: [], // Images are handled separately
    flowerLanguage: product.flowerLanguage || '',
    careGuide: product.careGuide || '',
    categoryId: product.categoryId || 0,
    specification: product.specification || '',
    status: product.status ?? 1,
    featured: product.featured ?? 0,
    stockQuantity: product.stockQuantity || 0,
    lowStockThreshold: product.lowStockThreshold || 5,
  };
};

// Validation helpers
export const validateProductForm = (data: unknown): {
  success: boolean;
  data?: ProductFormData;
  errors?: Record<string, string>;
} => {
  const result = productFormSchema.safeParse(data);

  if (result.success) {
    return {
      success: true,
      data: result.data,
    };
  }

  // Transform Zod errors to a more friendly format
  const errors: Record<string, string> = {};
  result.error.issues.forEach((issue) => {
    const path = issue.path.join('.');
    errors[path] = issue.message;
  });

  return {
    success: false,
    errors,
  };
};

File: flower_server/src/main/java/com/flower/shop/entity/Product.java
package com.flower.shop.entity;

import com.baomidou.mybatisplus.annotation.*;
import lombok.Data;
import lombok.EqualsAndHashCode;

import java.math.BigDecimal;
import java.time.LocalDateTime;

/**
 * 商品实体类
 *
 * 设计说明：
 * - 每个商品都关联花材分类和包装分类
 * - 支持多图片展示
 * - 包含价格、描述等基本信息
 */
@Data
@EqualsAndHashCode(callSuper = false)
@TableName("products")
public class Product {

    /**
     * 商品ID - 主键
     */
    @TableId(value = "id", type = IdType.AUTO)
    private Long id;

    /**
     * 商品名称
     * 格式建议：[花材名称] + [包装类型] + [特色描述]
     * 如：11朵红玫瑰花束、精美百合花篮
     */
    @TableField("name")
    private String name;

    /**
     * 花材分类ID（关联categories表，parent_id为NULL的分类）
     */
    @TableField("category_id")
    private Long categoryId;

    // 注意：当前数据库设计中没有packaging_id字段
// 包装信息通过category_id的层级关系来区分

    /**
     * 商品描述
     */
    @TableField("description")
    private String description;

    /**
     * 商品图片列表（JSON格式存储多个图片URL）
     * 数据库字段为：images (JSON类型)
     */
    @TableField("images")
    private String images;

    /**
     * 商品规格
     * 如：11朵、19朵、大号、中号、小号等
     */
    @TableField("specification")
    private String specification;

    /**
     * 当前售价
     */
    @TableField("price")
    private BigDecimal price;

    /**
     * 原价（用于显示折扣）
     */
    @TableField("original_price")
    private BigDecimal originalPrice;

    /**
     * 商品状态：0-下架，1-上架
     * 数据库字段为：status
     */
    @TableField("status")
    private Integer status;

    /**
     * 是否推荐：0-不推荐，1-推荐
     * 数据库字段为：featured
     */
    @TableField("featured")
    private Integer featured;

    /**
     * 花语说明
     * 数据库字段为：flower_language
     */
    @TableField("flower_language")
    private String flowerLanguage;

    /**
     * 养护指南
     * 数据库字段为：care_guide
     */
    @TableField("care_guide")
    private String careGuide;

    
    
    /**
     * 创建时间
     */
    @TableField(value = "created_at", fill = FieldFill.INSERT)
    private LocalDateTime createdAt;

    /**
     * 更新时间
     */
    @TableField(value = "updated_at", fill = FieldFill.INSERT_UPDATE)
    private LocalDateTime updatedAt;

    /**
     * 逻辑删除标记：0-未删除，1-已删除
     */
    @TableLogic
    @TableField("deleted")
    private Integer deleted;

    /**
     * 花材分类名称（临时存储，用于显示）
     */
    @TableField(exist = false)
    private String categoryName;

    /**
     * 库存数量
     */
    @TableField("stock_quantity")
    private Integer stockQuantity;

    /**
     * 低库存预警阈值
     */
    @TableField("low_stock_threshold")
    private Integer lowStockThreshold;

    /**
     * 判断是否有折扣
     */
    public boolean hasDiscount() {
        if (originalPrice == null || price == null) {
            return false;
        }
        return price.compareTo(originalPrice) < 0;
    }

    /**
     * 判断是否上架
     */
    public boolean isOnline() {
        return Integer.valueOf(1).equals(this.status);
    }

    /**
     * 判断是否推荐
     */
    public boolean isFeaturedProduct() {
        return Integer.valueOf(1).equals(this.featured);
    }

    /**
     * 判断是否库存不足
     */
    public boolean isLowStock() {
        if (stockQuantity == null || lowStockThreshold == null) {
            return false;
        }
        return stockQuantity > 0 && stockQuantity <= lowStockThreshold;
    }

    /**
     * 判断是否缺货
     */
    public boolean isOutOfStock() {
        return stockQuantity != null && stockQuantity <= 0;
    }

    /**
     * 判断库存充足
     */
    public boolean hasSufficientStock() {
        return stockQuantity != null && stockQuantity > lowStockThreshold;
    }

    /**
     * 获取库存状态文本
     */
    public String getStockStatusText() {
        if (isOutOfStock()) {
            return "缺货";
        } else if (isLowStock()) {
            return "库存不足";
        } else {
            return "库存充足";
        }
    }

    /**
     * 设置库存数量
     */
    public void setStockQuantity(Integer stockQuantity) {
        this.stockQuantity = stockQuantity;
        // 自动更新库存状态
        updateStockStatus();
    }

    /**
     * 设置低库存预警阈值
     */
    public void setLowStockThreshold(Integer lowStockThreshold) {
        this.lowStockThreshold = lowStockThreshold;
        // 自动更新库存状态
        updateStockStatus();
    }

    /**
     * 更新库存状态（内部方法）
     */
    private void updateStockStatus() {
        if (stockQuantity != null && lowStockThreshold != null) {
            // 可以在这里添加状态更新逻辑
        }
    }

    /**
     * 扣减库存
     */
    public void reduceStock(Integer quantity) {
        if (quantity != null && quantity > 0 && stockQuantity != null) {
            this.stockQuantity = Math.max(0, this.stockQuantity - quantity);
            updateStockStatus();
        }
    }

    /**
     * 增加库存
     */
    public void addStock(Integer quantity) {
        if (quantity != null && quantity > 0 && stockQuantity != null) {
            this.stockQuantity = stockQuantity + quantity;
            updateStockStatus();
        }
    }
}